unsource qa_crowbarsetup

jdsn opened the pull request:
Move qa_crowbarsetup help to common file.
Renaming `onadmin_help` to `qacrowbarsetup_help` which is a correct name (because the `onadmin_` prefix means that a function runs **on** the admin node).
    
The goal is to untie mkcloud and qa_crowbarsetup.
So we removed the sourcing of qa_crowbarsetup a few weeks ago (this is a feature).

The idea behind this is, that mkcloud can change more radical than qa_crowbarsetup. The latter will most likely remain a collection of shell functions. But mkcloud will need to change sooner or later into a OOP style (this will not be bash). Changes like this contradict this plan. Breakages should not be fixed by sourcing qa_crowbarsetup.sh again.

This PR contains the revert of commit dae0ddd21698e274175fee60c31d8ae8a24d5600

nicolasbock commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -12,6 +12,9 @@ for script in $common_scripts; do
 done
 
+# not being sourced from mkcloud is a feature
+is_onhost && complain 9 "qa_crowbarsetup.sh was sourced in mkcloud"
```
:+1:

aspiers replied:
Maybe the error message could be more helpful. If I hadn't seen this PR I would be totally confused by that.

nicolasbock commented on scripts/mkcloud:
```diff
@@ -1084,8 +1084,6 @@ Optional
         Full path to directory containing scenario file.
 EOUSAGE
-    # Source qa_crowbarsetup.sh, which is where onadmin_help lives
-    source ${SCRIPTS_DIR}/qa_crowbarsetup.sh
```
This commit breaks help, doesn't it? Could you squash this commit with the following one so that is doesn't?

s-t-e-v-e-n-k replied:
It won't break help, it will print mkcloud help and then print no file or directory for the function call immediately after it.

nicolasbock replied:
Well, that's kind of what I meant :smile: Printing 'no file or directory' for the help function breaks the mkcloud though. You can't call the help function anymore.

vmoravec commented:
@jdsn As QA and perhaps others still rely on sourcing qa_crowbarsetup.sh, what are the consequences for those using the file as standalone collection of functions?

jdsn commented:
> My only issue with this change is that developers have to remember to update two files if new flags get added.

@s-t-e-v-e-n-k 
Why? If new variables are added that are needed on the admin node then they have to be passed there in both cases. For all variables with the `want_` prefix this is done automatically. Others need a good reason to be named differently.

Remember, the goal is to reduce the number of variables that we need to configure an mkcloud run. We want to take as many configuration parameters from batch/scenario files as possible.

> I'd prefer the commits to be squashed.

@nicolasbock 
I agree that the revert alone reintroduces the failure, but the following commit from this PR contains the fix. For a normal change commit I would agree to squash, but in this case my preference is to have the revert as standalone commit - to show what was reverted, without mixing it. As the PR gets merged together anyway this should not be an issue.
But I don't care that much about it being one or two commits. So if this is preferred by all other parties, I can squash the commits (please speak up), otherwise I keep them separate.


> As QA and perhaps others still rely on sourcing qa_crowbarsetup.sh, what are the consequences for those using the file as standalone collection of functions?

@vmoravec This change only complains if qa_crowbarsetup was sourced within the mkcloud script. All other sourcing is unaffected. Even in our deployments qa_crowbarsetup is sourced in the admin node shell and the functions are called directly. So no changes here, and no changes for standalone usage.
In essence, this is exactly what the change enforces: that qa_crowbarsetup remains a standalone tool by not mixing it into mkcloud.

s-t-e-v-e-n-k commented:
@jdsn The variables themselves are not my concern -- my concern is that the variables are used in qa_crowbarsetup.sh, but have to be documented in lib/mkcloud-common.sh -- I'm just trying to point out that it would be easy for the code and the help to get out of sync. Perhaps a comment in qa_crowbarsetup.sh that points out where the help output is.

jdsn commented:
> ... have to be documented in lib/mkcloud-common.sh [...] Perhaps a comment in qa_crowbarsetup.sh that points out where the help output is.
@s-t-e-v-e-n-k 
Ah, now it makes sense. Sorry did not get it the first time.
Yes, I can add such a comment.

nicolasbock commented:
@jdsn Thanks. While it's true that the PR gets merged together, having commits in it that break the code somehow is potentially problematic when you try to bisect the repository. It just leads to additional bad commits which have no relevance to the bug you are trying to bisect. I am not saying that I have ever bisected `mkcloud` before or am planning on doing that in the near future. This is really more of a general remark :smile:.

aspiers commented:
@nicolasbock's point about introducing breakage in the history is spot on.  Normally I would agree that no commit should *ever* intentionally introduce breakage.  (Of course we are all human, so occasional unintentional breakage is inevitable.)  Having said that, I can see @jdsn's point that keeping the revert as a separate commit also makes sense, so I'm conflicted right now ;-)  One could argue that the revert is deliberately *re-*introducing breakage in order to fix it, so maybe that's not as bad ...  Probably not worth debating for hours though ;-)

nicolasbock commented:
@aspiers I share your conflictedness :) But @jdsn convinced me nicely that it makes more sense in this case to leave it as two commits.

aspiers commented on scripts/lib/mkcloud-common.sh:
```diff
@@ -539,3 +539,42 @@ function dist_to_image_name
 
 # ---- END: functions related to repos and distribution settings
+
+function qacrowbarsetup_help
```
If this is help specific to `qacrowbarsetup.sh` then why is it in a file called `mkcloud-common.sh`? Wouldn't `qacrowbarsetup-help.sh` or `qacrowbarsetup-shared.sh` make more sense?

jdsn replied:
@aspiers 
Agreed, the message should be more specific.

Yes, the name `mkcloud-common.sh` is not the best. This file contains functions that are common to `mkcloud` and `qa_crowbarsetup.sh`. Maybe `mkc-qac-common.sh` or `m-q-common.sh` would be better. But having a separate file like `qacrowbarsetup-shared.sh` is also a nice idea. It's yet another file, but would also make sense.

jdsn commented:
Extracted the fix for the onhost detection: now in https://github.com/SUSE-Cloud/automation/pull/1421

The help is now moved to a separate file and the error message is more specific as proposed from @aspiers .

aspiers commented on scripts/qa_crowbarsetup.sh:
```diff
@@ -7,11 +7,14 @@ shopt -s extglob
 done
 
+# not being sourced from mkcloud is a feature
+is_onhost && complain 9 "qa_crowbarsetup.sh should not be sourced within mkcloud. Use the common files for shared functions."
```
This is still cryptic for someone who doesn't understand the issue.  Put yourself in their shoes :)  They will ask "Which common files?  Which shared functions?  What are you talking about?  How the hell do I fix this?"

IIUC, this can only happen if an old mkcloud accidentally uses a newer `qa_crowbarsetup.sh`, right?  So a much more helpful message would be to simply explain that.  Then they can easily understand that they need to update their `mkcloud`.

jdsn replied:
> this can only happen if an old mkcloud accidentally uses a newer `qa_crowbarsetup.sh`

@aspiers 
Thats one part. The other goal is to document this as a feature in order to let PR gatings fail that reintroduce the sourcing later (like the commit that I reverted in this PR).

> This is still cryptic for someone who doesn't understand the issue.

Would it be acceptable to solve this by documenting the current way the mkcloud files are used in a few lines at the top of the file (also describing the common files) and here pointing to that place?
